File: frontend/src/api/tasks.ts
import { apiClient } from './client';

export interface TaskPlanStep {
  step_number: number;
  file: string;
  symbol: string;
  operation: string;
  reason: string;
  risk: string;
  expected_result: string;
}

export interface RootCauseEvidence {
  confidence: string;
  summary: string;
  culprit_file?: string;
  culprit_symbol?: string;
  evidence_points: string[];
  call_chain: string[];
  related_tests: string[];
}

export interface EngineeringTask {
  task_id: string;
  title: string;
  description: string;
  project_id: string;
  project_root: string;
  status: string;
  priority: string;
  task_type: string;
  target_files: string[];
  target_symbols: string[];
  discovered_symbols: string[];
  affected_files: string[];
  root_cause?: RootCauseEvidence;
  impact: Record<string, any>;
  implementation_plan: TaskPlanStep[];
  patch?: string;
  proposal_id?: string;
  tests_discovered: string[];
  tests_generated: string[];
  validation_results: Record<string, any>;
  review_results: Record<string, any>;
  pr_summary?: string;
  risk: string;
  iteration_count: number;
  max_iterations: number;
  checkpoint_id?: string;
  error_message?: string;
  decision_reason?: string;
  created_at: string;
  updated_at: string;
}

export interface TaskListResponse {
  success: boolean;
  total: number;
  tasks: EngineeringTask[];
}

export interface TaskResponse {
  success: boolean;
  task: EngineeringTask;
}

export const tasksApi = {
  createTask: (title: string, description: string = '', task_type?: string, priority?: string, project_id?: string) =>
    apiClient<TaskResponse>('/tasks', {
      method: 'POST',
      body: JSON.stringify({ title, description, task_type, priority, project_id }),
    }),

  listTasks: (status?: string, task_type?: string, priority?: string) =>
    apiClient<TaskListResponse>('/tasks', {
      params: { status, task_type, priority },
    }),

  getTask: (taskId: string) =>
    apiClient<TaskResponse>(`/tasks/${taskId}`),

  analyzeTask: (taskId: string) =>
    apiClient<TaskResponse>(`/tasks/${taskId}/analyze`, {
      method: 'POST',
    }),

  planTask: (taskId: string) =>
    apiClient<TaskResponse>(`/tasks/${taskId}/plan`, {
      method: 'POST',
    }),

  approveTask: (taskId: string, reason?: string, force: boolean = false) =>
    apiClient<TaskResponse>(`/tasks/${taskId}/approve`, {
      method: 'POST',
      body: JSON.stringify({ reason, force }),
    }),

  rejectTask: (taskId: string, reason?: string) =>
    apiClient<TaskResponse>(`/tasks/${taskId}/reject`, {
      method: 'POST',
      body: JSON.stringify({ reason }),
    }),

  executeTask: (taskId: string, runTests: boolean = true) =>
    apiClient<TaskResponse>(`/tasks/${taskId}/execute`, {
      method: 'POST',
      body: JSON.stringify({ run_tests: runTests }),
    }),

  rollbackTask: (taskId: string) =>
    apiClient<TaskResponse>(`/tasks/${taskId}/rollback`, {
      method: 'POST',
    }),

  getTaskDiff: (taskId: string) =>
    apiClient<{ success: boolean; task_id: string; patch: string }>(`/tasks/${taskId}/diff`),

  getTaskReport: (taskId: string) =>
    apiClient<{ success: boolean; task_id: string; title: string; status: string; pr_summary: string }>(`/tasks/${taskId}/report`),
};

File: app/api/tasks.py
"""
DevPilot Task Management REST API Endpoints (v3.4).

Exposes autonomous issue-to-PR task lifecycle endpoints:
creation, analysis, planning, approval, rejection, execution, test verification,
diff inspection, rollback, and PR report generation.
"""

from pathlib import Path
from typing import List, Optional
from fastapi import APIRouter, HTTPException, Query, status

from app.schemas.tasks import (
    ApproveTaskRequest,
    CreateTaskRequest,
    ExecuteTaskRequest,
    RejectTaskRequest,
    TaskListResponse,
    TaskReportResponse,
    TaskResponse,
)
from app.tasks.engine import EngineeringTaskEngine
from app.tasks.models import InvalidTaskStateTransitionError
from app.tasks.store import TaskStore

router = APIRouter(tags=["Task Engineering (v3.4)"])


def _get_engine(project_root: Optional[str] = None) -> EngineeringTaskEngine:
    root = Path(project_root).resolve() if project_root else Path.cwd().resolve()
    return EngineeringTaskEngine(project_root=root)


def _get_store(project_root: Optional[str] = None) -> TaskStore:
    root = Path(project_root).resolve() if project_root else Path.cwd().resolve()
    return TaskStore(project_root=root)


@router.post(
    "/tasks",
    response_model=TaskResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Create Engineering Task",
    description="Initializes a new issue-to-PR autonomous engineering task.",
)
def create_task(req: CreateTaskRequest) -> TaskResponse:
    engine = _get_engine()
    task = engine.create_task(
        title=req.title,
        description=req.description or "",
        task_type=req.task_type,
        priority=req.priority,
        project_id=req.project_id or "default",
    )
    return TaskResponse(task=task.to_dict())


@router.get(
    "/tasks",
    response_model=TaskListResponse,
    summary="List Engineering Tasks",
    description="Lists existing engineering tasks with optional status and priority filters.",
)
def list_tasks(
    status: Optional[str] = Query(None, description="Filter by status (CREATED, ANALYZED, PLANNED, APPROVED, COMPLETED, etc.)"),
    task_type: Optional[str] = Query(None, description="Filter by task type"),
    priority: Optional[str] = Query(None, description="Filter by priority"),
) -> TaskListResponse:
    store = _get_store()
    tasks = store.list_tasks(status=status, task_type=task_type, priority=priority)
    return TaskListResponse(total=len(tasks), tasks=[t.to_dict() for t in tasks])


@router.get(
    "/tasks/{task_id}",
    response_model=TaskResponse,
    summary="Get Engineering Task",
    description="Retrieves full task details, analysis, root cause, plan, and PR summary.",
)
def get_task(task_id: str) -> TaskResponse:
    store = _get_store()
    task = store.get(task_id)
    if not task:
        raise HTTPException(status_code=404, detail=f"Task '{task_id}' not found.")
    return TaskResponse(task=task.to_dict())


@router.post(
    "/tasks/{task_id}/analyze",
    response_model=TaskResponse,
    summary="Analyze Task & Root Cause",
    description="Performs codebase analysis, symbol resolution, and root cause evidence discovery.",
)
def analyze_task(task_id: str) -> TaskResponse:
    engine = _get_engine()
    try:
        task = engine.analyze_task(task_id)
        return TaskResponse(task=task.to_dict())
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except InvalidTaskStateTransitionError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Analysis failed: {str(e)}")


@router.post(
    "/tasks/{task_id}/plan",
    response_model=TaskResponse,
    summary="Generate Implementation Plan & Patch",
    description="Constructs step-by-step implementation plan and reviewable patch proposal.",
)
def plan_task(task_id: str) -> TaskResponse:
    engine = _get_engine()
    try:
        task = engine.plan_task(task_id)
        return TaskResponse(task=task.to_dict())
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except InvalidTaskStateTransitionError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Planning failed: {str(e)}")


@router.post(
    "/tasks/{task_id}/approve",
    response_model=TaskResponse,
    summary="Approve Engineering Task",
    description="Approves task and authorizes safe patch execution.",
)
def approve_task(task_id: str, req: ApproveTaskRequest) -> TaskResponse:
    engine = _get_engine()
    try:
        task = engine.approve_task(task_id, reason=req.reason, force=req.force)
        return TaskResponse(task=task.to_dict())
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except InvalidTaskStateTransitionError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Approval failed: {str(e)}")


@router.post(
    "/tasks/{task_id}/reject",
    response_model=TaskResponse,
    summary="Reject Engineering Task",
    description="Rejects task and records developer decision.",
)
def reject_task(task_id: str, req: RejectTaskRequest) -> TaskResponse:
    engine = _get_engine()
    try:
        task = engine.reject_task(task_id, reason=req.reason)
        return TaskResponse(task=task.to_dict())
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except InvalidTaskStateTransitionError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Rejection failed: {str(e)}")


@router.post(
    "/tasks/{task_id}/execute",
    response_model=TaskResponse,
    summary="Execute Approved Task",
    description="Executes approved task with safety checkpoints, syntax verification, test verification, and post-review.",
)
def execute_task(task_id: str, req: ExecuteTaskRequest) -> TaskResponse:
    engine = _get_engine()
    try:
        task = engine.execute_task(task_id, run_tests=req.run_tests)
        return TaskResponse(task=task.to_dict())
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except InvalidTaskStateTransitionError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Execution failed: {str(e)}")


@router.post(
    "/tasks/{task_id}/rollback",
    response_model=TaskResponse,
    summary="Rollback Task",
    description="Reverts task changes from backup checkpoint.",
)
def rollback_task(task_id: str) -> TaskResponse:
    engine = _get_engine()
    try:
        task = engine.rollback_task(task_id)
        return TaskResponse(task=task.to_dict())
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except InvalidTaskStateTransitionError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Rollback failed: {str(e)}")


@router.get(
    "/tasks/{task_id}/diff",
    summary="Get Task Diff",
    description="Retrieves the proposed or executed unified diff for a task.",
)
def get_task_diff(task_id: str):
    store = _get_store()
    task = store.get(task_id)
    if not task:
        raise HTTPException(status_code=404, detail=f"Task '{task_id}' not found.")
    return {"success": True, "task_id": task_id, "patch": task.patch or ""}


@router.get(
    "/tasks/{task_id}/report",
    response_model=TaskReportResponse,
    summary="Get Task PR Report",
    description="Retrieves the PR-ready markdown package for a task.",
)
def get_task_report(task_id: str) -> TaskReportResponse:
    store = _get_store()
    task = store.get(task_id)
    if not task:
        raise HTTPException(status_code=404, detail=f"Task '{task_id}' not found.")
    return TaskReportResponse(
        task_id=task.task_id,
        title=task.title,
        status=task.status,
        pr_summary=task.pr_summary or "(PR summary pending completion)",
    )
